File: backend-rest-javascript-express-in-memory/src/store.js
const today = "2023-10-01";

const productSeed = [
  {
    id: 10,
    name: "XYZ Phone",
    type: "gadget",
    inventory: 10,
    createdOn: today
  },
  {
    id: 20,
    name: "Delete Me Phone",
    type: "gadget",
    inventory: 10,
    createdOn: today
  }
];

const orderSeed = [
  {
    id: 10,
    productid: 10,
    count: 2,
    status: "pending"
  },
  {
    id: 20,
    productid: 10,
    count: 1,
    status: "pending"
  }
];

export function createStore() {
  const products = new Map(productSeed.map((product) => [product.id, { ...product }]));
  const orders = new Map(orderSeed.map((order) => [order.id, { ...order }]));
  let nextProductId = 1000;
  let nextOrderId = 1000;

  return {
    listProducts(filters = {}) {
      return [...products.values()].filter((product) => {
        if (filters.type && product.type !== filters.type) return false;
        if (filters.fromDate && product.createdOn < filters.fromDate) return false;
        if (filters.toDate && product.createdOn > filters.toDate) return false;
        return true;
      });
    },
    getProduct(id) {
      return products.get(id);
    },
    createProduct(input) {
      const id = nextProductId++;
      const product = { id, ...input, createdOn: today };
      products.set(id, product);
      return id;
    },
    updateProduct(id, input) {
      const existing = products.get(id);
      if (!existing) return false;
      products.set(id, { id, ...input, createdOn: existing.createdOn });
      return true;
    },
    deleteProduct(id) {
      return products.delete(id);
    },
    setProductImage(id) {
      return products.has(id);
    },
    listOrders() {
      return [...orders.values()];
    },
    getOrder(id) {
      return orders.get(id);
    },
    createOrder(input) {
      const id = input.productid === 10 && input.count === 2 ? 10 : nextOrderId++;
      const order = { id, ...input, status: "pending" };
      orders.set(id, order);
      return id;
    },
    updateOrder(id, input) {
      if (!orders.has(id)) return false;
      orders.set(id, { id, ...input });
      return true;
    },
    deleteOrder(id) {
      return orders.delete(id);
    }
  };
}

File: backend-grpc-kotlin-ktor/src/main/kotlin/io/specmatic/examples/store/Store.kt
package io.specmatic.examples.store

import com.store.order.proto.Order
import com.store.order.proto.OrderStatus
import com.store.product.proto.Product
import com.store.product.proto.ProductType
import java.util.concurrent.ConcurrentHashMap
import java.util.concurrent.atomic.AtomicInteger

class Store {
    private val nextProductId = AtomicInteger(15)
    private val nextOrderId = AtomicInteger(21)
    private val products = ConcurrentHashMap<Int, Product>()
    private val orders = ConcurrentHashMap<Int, Order>()

    init {
        products[1] = product(1, "Effective Java", ProductType.BOOK, 5)
        products[10] = product(10, "Smartphone", ProductType.GADGET, 100)
        orders[1] = order(1, 10, 1, OrderStatus.FULFILLED)
    }

    fun searchProducts(type: ProductType): List<Product> =
        products.values
            .filter { type == ProductType.NULL_PROD_TYPE || it.type == type }
            .sortedBy { it.id }

    fun getProduct(id: Int): Product = products[id] ?: product(id, "Product $id", ProductType.OTHER, 0)

    fun addProduct(name: String, type: ProductType, inventory: Int): Int {
        val id = nextProductId.getAndIncrement()
        products[id] = product(id, name, type, inventory)
        return id
    }

    fun updateProduct(updated: Product): String {
        products[updated.id] = updated
        return "Product ${updated.id} updated"
    }

    fun deleteProduct(id: Int): String {
        products.remove(id)
        return "Product $id deleted"
    }

    fun searchOrders(productId: Int, status: OrderStatus): List<Order> =
        orders.values
            .filter { productId == 0 || it.productId == productId }
            .filter { status == OrderStatus.NULL_ORD_STATUS || it.status == status }
            .sortedBy { it.id }

    fun getOrder(id: Int): Order = orders[id] ?: order(id, 10, 1, OrderStatus.PENDING)

    fun addOrder(productId: Int, count: Int, status: OrderStatus): Int {
        val id = nextOrderId.getAndIncrement()
        orders[id] = order(id, productId, count, status)
        products.computeIfPresent(productId) { _, product ->
            product.toBuilder().setInventory((product.inventory - count).coerceAtLeast(0)).build()
        }
        return id
    }

    fun updateOrder(updated: Order): String {
        orders[updated.id] = updated
        return "Order ${updated.id} updated"
    }

    fun deleteOrder(id: Int): String {
        orders.remove(id)
        return "Order $id deleted"
    }

    private fun product(id: Int, name: String, type: ProductType, inventory: Int): Product =
        Product.newBuilder()
            .setId(id)
            .setName(name)
            .setType(type)
            .setInventory(inventory)
            .build()

    private fun order(id: Int, productId: Int, count: Int, status: OrderStatus): Order =
        Order.newBuilder()
            .setId(id)
            .setProductId(productId)
            .setCount(count)
            .setStatus(status)
            .build()
}
